attribute-service-impl/src/test/java/org/hypertrace/core/attribute/service/model/AttributeMetadataModelTest.java

```diff
@@ -256,4 +253,108 @@ public void testAttributeDefinitionBackwardsCompatibility() throws IOException {
+    AttributeMetadata expected =
+        template.toBuilder()
+            .setScope(AttributeScope.EVENT)
+            .setScopeString("EVENT")
+            .setId("EVENT.key")
+            .build();
+    Assertions.assertEquals(expected, AttributeMetadataModel.fromDTO(input).toDTO());
+    Assertions.assertEquals(
+        AttributeMetadataModel.fromDTO(expected).toJson(),
+        AttributeMetadataModel.fromDTO(input).toJson());
+
+    // String scope backwards compatible when possible
+    input = template.toBuilder().setScopeString("EVENT").build();
+    expected =
+        template.toBuilder()
+            .setScope(AttributeScope.EVENT)
+            .setScopeString("EVENT")
+            .setId("EVENT.key")
+            .build();
+    Assertions.assertEquals(expected, AttributeMetadataModel.fromDTO(input).toDTO());
+    Assertions.assertEquals(
+        AttributeMetadataModel.fromDTO(expected).toJson(),
+        AttributeMetadataModel.fromDTO(input).toJson());
+
+    // If scope string not representable by enum, enum set to undefined and string used
+    input = template.toBuilder().setScopeString("NEWSCOPE").build();
+    expected =
+        template.toBuilder()
+            .setScopeString("NEWSCOPE")
+            .setId("NEWSCOPE.key")
+            .build();
+    Assertions.assertEquals(expected, AttributeMetadataModel.fromDTO(input).toDTO());
+    Assertions.assertEquals(
+        AttributeMetadataModel.fromDTO(expected).toJson(),
+        AttributeMetadataModel.fromDTO(input).toJson());
+
+    // Given Json with scope, scope and scope string should be set
+    String inputJson =
+        "{"
+            + "\"fqn\":\"fqn\","
+            + "\"key\":\"key\","
+            + "\"scope\":\"EVENT\","
+            + "\"type\":\"ATTRIBUTE\","
+            + "\"value_kind\":\"TYPE_BOOL\","
+            + "\"display_name\":\"Some Name\""
+            + "}";
+    expected =
+        template.toBuilder()
+            .setScope(AttributeScope.EVENT)
+            .setScopeString("EVENT")
+            .setId("EVENT.key")
+            .build();
+    Assertions.assertEquals(expected, AttributeMetadataModel.fromJson(inputJson).toDTO());
+
+    // Given JSON with scope string, scope and scope string should be set
+    inputJson =
+        "{"
+            + "\"fqn\":\"fqn\","
+            + "\"key\":\"key\","
+            + "\"scope_string\":\"EVENT\","
+            + "\"type\":\"ATTRIBUTE\","
+            + "\"value_kind\":\"TYPE_BOOL\","
+            + "\"display_name\":\"Some Name\""
+            + "}";
+    expected =
+        template.toBuilder()
+                .setScope(AttributeScope.EVENT)
+                .setScopeString("EVENT")
+                .setId("EVENT.key")
+                .build();
+    Assertions.assertEquals(expected, AttributeMetadataModel.fromJson(inputJson).toDTO());
+
+    // Given JSON with unknown scope string, scope string only should be set
```

Comment:
what about both scope and scope_string? is that an important path?

Reply:
I'd guess the one defined later would win. It definitely wouldn't break. 
Regardless, it's not an important path - the json schema is changed with this change - scope can be read from the database but no longer written (ln 155 of AttributeMetadataMode). So existing entries would read a scope and set both values of an attribute (case on line 325 here), anything being written (or updated) after this change would write only scope_string.

Reply:
ok